### ISSUE TITLE
Tag new release

### ISSUE THREAD

**Issue by KyleSanderson** [hasse69/rar2fs#50]
Hi Hasse69,

When https://github.com/hasse69/rar2fs/issues/49 is resolved or https://github.com/hasse69/rar2fs/issues/47 (or sooner, if you feel like it), could you please tag the release so distro maintainers can get atleast the ISO fix? I'm seeing sporadic corruption recently with rar2fs; not sure what's going on.

**Comment by hasse69** [comment]
I can prepare a new release sometime next week.
But nothing fixed between last release and current version should address any stability issues. If you experience problems like the one you describe you need to provide more details.  Corruption on file system level is very unlikely since rat2fs does not allow modifications to the source files.

**Comment by KyleSanderson** [comment]
I'm seeing corruption with rar2fs where mpc-hc using essentially rarfilesource is playing the files back no problem. I've been seeing serious perf issues with directory listings/etc so I'm hoping caching takes care of it. However if the iso stuff doesn't resolve the issue then can I provide more data.

**Comment by hasse69** [comment]
Are you having problems with Iso playback? Otherwise I do not see any need for the patch.
Performance wise I do not know of any issues with directory listing etc except for encrypted archives which currently does not work at all. Are you using the --seek-length option? If not, try that. Caching is a secondary approach to speeding up the process and initially the performance penalty is the same.
Using e.g. --seek-length=1 should be the first thing to try.
So, please provide some details about what files rar2fs cannot handle compared to rarfilesource.
And also when and if this was working before.

**Comment by hasse69** [comment]
I would like to complete issue #49 before I tag a new release.
Any news on the performance issues?

**Comment by KyleSanderson** [comment]
Yeah, sorry for the delay on this.

Corrupt archives really, missing pieces. Rarfilesource seems to just play them then die, whereas rar2fs seems to just bail?

fuse.rar2fs (ro,nosuid,nodev,relatime,user_id=0,group_id=0,default_permissions,allow_other)


/mnt/Recordings                                         /mnt/RecordingsU                        rar2fs          ro,allow_other,uid=kyle,gid=storage,umask=0222,--seek-length=2  0 0

I had --seek-length=2 in my /etc/fstab, evidently this did not actually get passed to mount? Any tips?


Reading the dir over smb takes a couple minutes (with tons of rars in subdirectories). Whereas using the native ext4 mount it's near instant. Directory refreshes are marginally faster, but that's it. 

This is with 385 directories.

**Comment by hasse69** [comment]
For the fstab issue, try a different syntax:

  rar2fs#/mnt/Recordings /mnt/RecordingsU fuse ro,allow_other,uid=kyle,gid=storage,umask=0222,--seek-length=2 0 0

Have not heard (nor observed it myself) that options would not take effect through fstab mounts before. Please try the variant above and report back.

Regarding the SMB performance, I do not really see how this can be solved by improving anything in rar2fs? To me it sounds more like a SMB bottle neck? As you stated, when testing natively it works ok? From rar2fs point of view it cannot tell if the mount point is exported over SMB and any operations on the rar2fs file system is the same with or without it. If the queries are delay due to network latency so will the observed "performance".

Can you explain a bit more about what you mean when you say that rar2fs bails on incomplete archives? Also, are we talking compressed archives or archives in store mode? These two types are handled a bit differently so it is important for me to know which of the two (or both) you believe are causing the problem. I assume playback is never started, right? But does the file appear on the file system?

**Comment by KyleSanderson** [comment]
```rar2fs on /mnt/RecordingsU type fuse.rar2fs (ro,relatime,user_id=0,group_id=0,default_permissions,allow_other)
```
is the result from
```
rar2fs#/mnt/Recordings                                          /mnt/RecordingsU                        fuse            ro,allow_other,uid=kyle,gid=storage,umask=0222,--seek-length=2  0 0
```

With SMB the directory listings are instant without rar2fs (I made the other share when I realized I couldn't play some files back). With the rar2fs share, the total listing takes a couple minutes in comparison to under 2 seconds. I'm unsure if this is because of fuse, or rar2fs. rar2fs is only using 10% of a core while the directory is listing out over SMB (ironically, more has to be networked without fuse, because each file is shown too).

It could just be the player is more reliable when using the filter... The difference is the .rar file is missing, with just .r00 being present going until the end. Player claims the track is 11h (with rar2fs) instead of the normal 56m~ (with rarfilesource).

**Comment by KyleSanderson** [comment]
Sorry for double posting, edited the above with proper syntax and a bit more info.

Rebooted the box, seeing the same performance with rar2fs versus mount.rar2fs (which was the previous invocation/syntax).

```
root      3934  0.1  0.0  95972  2696 ?        Ssl  15:58   0:00 rar2fs /mnt/Web /mnt/WebU -o ro,allow_other,uid=1000,gid=1001,umask=0222,--seek-length=2,dev,suid
root      3938  0.2  0.1  96004  3756 ?        Ssl  15:58   0:00 rar2fs /mnt/Backup /mnt/BackupU -o ro,allow_other,uid=1000,gid=1001,umask=0222,--seek-length=2,dev,suid
root      3942  0.0  0.0  95960   736 ?        Ssl  15:58   0:00 rar2fs /mnt/Recordings /mnt/RecordingsU -o ro,allow_other,uid=1000,gid=1001,umask=0222,--seek-length=2,dev,suid
```

**Comment by hasse69** [comment]
The fstab works just fine, I think you might assume that all options specified in fstab like --seek-length=2 would appear in the output of 'mount' which is actually not correct. The options that are specific to rar2fs is not presented by mount since they are all terminated in the user space application and not part of the actual mount subsystem (including the fuse driver in kernel).
Try setting seek length to a much higher value. I think you might notice a really huge difference in performance, and not in a good way.

I cannot explain why SMB exports of a rar2fs file system would perform any worse than a local rar2fs mount? At least I would not expect it to be slower than what is the difference between native and rar2fs mounts locally. If it takes 2 min. using SMB I would expect a local rar2fs mount to also take 2 min. But then it also depends on how SMB works. Maybe it does recursive lookups and then you must compare using e.g. ls -lR locally, which of course is going to be magnitudes slower,

Regarding the missing files. I still have a bit of hard time understanding the use case :(
Are you missing some of the volume files in the middle somewhere?

**Comment by KyleSanderson** [comment]
SMB does indeed do a recursive lookup.

The comment was just that rarfilesource seems to handle missing archives much better, whereas rar2fs seems to give bad data to the player. As to what this means code wise is beyond me as I haven't compared the two. Trying to get perf top working so I can atleast tell you which functions are taking eons.

```
kyle@FileServer /mnt/Backup/Stuff/Auto $ time ls -la | wc -l
574

real    0m0.061s
user    0m0.010s
sys     0m0.027s
kyle@FileServer /mnt/Backup/Stuff/Auto $ time ls -lRa | wc -l
28326

real    0m1.126s
user    0m0.570s
sys     0m0.580s
kyle@FileServer /mnt/Backup/Stuff/Auto $ cd -
/mnt/BackupU/Stuff/Auto
kyle@FileServer /mnt/BackupU/Stuff/Auto $ time ls -la | wc -l
574

real    0m0.180s
user    0m0.017s
sys     0m0.067s
kyle@FileServer /mnt/BackupU/Stuff/Auto $ time ls -lRa | wc -l
10572

real    0m33.002s
user    0m0.270s
sys     0m0.843s
kyle@FileServer /mnt/BackupU/Stuff/Auto $
```

**Comment by hasse69** [comment]
Did you see my edits in previous post? Could you comment on at least the last one.

**Comment by KyleSanderson** [comment]
If files are missing in the middle, the player stops when it encounters the missing archives. rarfilesource will play through them resulting in visual, corruption for the duration the files are missing. If the initial archive (the .rar) is missing then the beginning will be missing but that's totally acceptable because the remaining files are present.

rar2fs seems to kill the player when an archive is outstanding.
rarfilesource continues along when an archive is outstanding.

**Comment by hasse69** [comment]
Ok. So what scenario is causing your player to report a bogus time?
Missing first .rar, missing volume file(s) in the middle, missing last volume file(s)?

I think it is a matter of how choice really. Instead of trying to extract incomplete archives, rar2fs is defensive and aborts. Would be ok to extract if it was media playback perhaps, but there are plenty of other cases that should never allow that. But if you can answer the question above I might have some time to look into it. I prefer a consistent/deterministic default application behavior, allowing extraction of incomplete/corrupt archives would be the opposite of that. But an option to telll rar2fs it should try might be possible to add. But I need to look more into the details before I can say for sure it will happen.

**Comment by KyleSanderson** [comment]
Seems like my portion unfortunately. Confirmed with the beginning, if something is missing in the middle the player won't open the file at all.

**Comment by hasse69** [comment]
Interesting. So you are saying that if files are missing in the beginning rar2fs will allow it but report wrong information (actually it does not report anything, time is read from the media file index by the player). But, if files are missing in the middle it will not play at all? 

Ok, that could make sense (even if it does not sound like it), but time is usually extracted from the end of the file and if the beginning is missing rar2fs probably starts from the first existing file and then ends up in the wrong place due to the offset error. If however a files is missing in the middle rar2fs will abort with an IO error since it cannot open the file.

**Comment by hasse69** [comment]
Btw, I think you can make rarfilesource to fail too, by removing volume files starting from the end. Then the player would not be able to read the index information.

**Comment by KyleSanderson** [comment]
The time is wrecked, audio is sped up, no picture. This is with mp4 in an mkv.

**Comment by hasse69** [comment]
This was with rarfilesource?

**Comment by KyleSanderson** [comment]
Sorry. That issue is with rar2fs. rarfilesource is a champ and plays the file anyways, correct time, just missing beginning frames.

**Comment by hasse69** [comment]
Gotta hand it over to them then. Nice work! So rarfilesource can return something that does not exist ;) I am afraid I do not posses that kind of magic :(

**Comment by hasse69** [comment]
There is a now a new tagged and released version 1.23.1.
I consider this issue as closed.
